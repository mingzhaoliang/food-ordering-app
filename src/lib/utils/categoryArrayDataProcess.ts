import { FieldValues, Path, UseFormReturn } from "react-hook-form";
import { ZodIssue } from "zod";

export interface FieldValue {
  name: string;
  image: { publicId: string; type: string; resourceType: string };
  imageFile: File;
}

const escapeRegExp = (string: string): string => {
  // Escapes special characters
  return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
};

const transformToFormData = <T extends FieldValues>(
  form: UseFormReturn<T, any, undefined>,
  formData: FormData,
  name: Path<T>
) => {
  (form.getValues(name) as FieldValue[]).forEach((item, index) => {
    formData.set(`${name}.${index}.name`, item.name);
    formData.set(`${name}.${index}.image`, JSON.stringify(item.image));
    formData.append(`${name}.${index}.imageFile`, item.imageFile);
  });
};

const transformFromFormData = (data: Record<string, string | object | File>, name: string) => {
  const transformedField: FieldValue[] = [];
  const escapedName = escapeRegExp(name);
  const regex = new RegExp(`^${escapedName}\\.(\\d+)\\.(\\w+)$`);

  Object.keys(data)
    .filter((key) => key.startsWith(`${name}.`))
    .forEach((key) => {
      // Extract the index and the field name (e.g., "0" and "name")
      const match = key.match(regex);
      if (match) {
        const index = Number(match[1]);
        const field = match[2] as keyof FieldValue;

        // Ensure the array has enough slots
        transformedField[index] = transformedField[index] || {};

        // Assign the field value to the corresponding object
        transformedField[index][field] = field === "image" ? JSON.parse(data[key] as string) : data[key];
      }

      // Remove the original flattened key from the transformed data
      delete data[key];
    });

  data[name] = transformedField;
};

const transformErrors = (errors: ZodIssue[], name: string) => {
  const transformedErrors = errors
    .filter((error) => error.path.includes(name))
    .reduce((acc, error) => {
      const index = Number(error.path[1]);
      acc[index] = acc[index] || {};
      acc[index][error.path[2]] = [error.message];
      return acc;
    }, [] as Array<Record<string, string[]>>);

  return transformedErrors;
};

export { transformErrors, transformFromFormData, transformToFormData };
