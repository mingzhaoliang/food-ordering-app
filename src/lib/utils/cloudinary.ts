import { CloudinaryIdentifier } from "@/types/CloudinaryIdentifier";

const cloudinaryUrl = ({ publicId, type, resourceType, version }: CloudinaryIdentifier, transformation?: string) => {
  return (
    "https://res.cloudinary.com/" +
    process.env.NEXT_PUBLIC_CLOUDINARY_CLOUD_NAME +
    "/" +
    resourceType +
    "/" +
    type +
    "/" +
    (version ? `v${version}` + "/" : "") +
    (transformation ? transformation + "/" : "") +
    publicId
  );
};

const cloudinaryIdentifier = (uploadResponse?: Record<string, any>) => {
  if (!uploadResponse) {
    return { publicId: "", type: "", resourceType: "", version: "" };
  }
  const { public_id, type, resource_type, version } = uploadResponse;
  return { publicId: public_id, type, resourceType: resource_type, version };
};

export { cloudinaryIdentifier, cloudinaryUrl };
