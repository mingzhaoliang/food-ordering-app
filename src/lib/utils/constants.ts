const OPENING_HOURS = [
  { day: "Sun - Mon", hours: ["closed"] },
  { day: "Tue - Thur", hours: ["6 pm - 9 pm"] },
  { day: "Fri - Sat", hours: ["12 pm - 10 pm"] },
];

const AU_STATES = ["ACT", "NSW", "NT", "QLD", "SA", "TAS", "VIC", "WA"] as const;
const AU_STATES_FULL = [
  "Australian Capital Territory",
  "New South Wales",
  "Northern Territory",
  "Queensland",
  "South Australia",
  "Tasmania",
  "Victoria",
  "Western Australia",
] as const;

const PAGES = [
  { href: "/", text: "Home" },
  { href: "/menu", text: "Menu" },
  { href: "/about", text: "About" },
  { href: "/contact", text: "Contact" },
];

const MAX_FILE_SIZE = 3 * 1024 * 1024; // 3MB
const ACCEPTED_IMAGE_TYPES = ["image/jpeg", "image/jpg", "image/png", "image/webp"];

export { ACCEPTED_IMAGE_TYPES, AU_STATES, AU_STATES_FULL, MAX_FILE_SIZE, OPENING_HOURS, PAGES };
