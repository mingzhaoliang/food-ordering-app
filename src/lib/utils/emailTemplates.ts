const verificationEmailTemplate = (restaurantName: string, userFirstName: string, verificationCode: string) => `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Email Verification</title>
    <style>
      body {
        font-family: Arial, sans-serif;
        margin: 0;
        padding: 0;
      }
      .container {
        width: 100%;
        max-width: 30rem;
        margin: 0 auto;
        overflow: hidden;
      }
      .header {
        padding: 0.5rem;
        text-align: center;
      }
      .header h2 {
        color: #737373;
      }
      .content {
        padding: 0.5rem;
        line-height: 1.6;
        text-align: center;
        text-wrap: balance;
        color: #737373;
        font-size: 1rem;
      }
      .code {
        display: inline-block;
        padding: 10px 15px;
        color: black;
        border-radius: 5px;
        font-size: 30px;
        margin: 1rem 0;
        font-weight: bold;
        letter-spacing: 0.05rem;
      }
      .footer {
        text-align: center;
        padding: 10px;
        font-size: 12px;
        color: #777;
        border-top: 1px solid #e0e0e0;
      }
    </style>
  </head>
  <body>
    <div class="container">
      <div class="header">
        <h2>${restaurantName}</h2>
        <h1>Verify your email</h1>
      </div>
      <div class="content">
        <p>Hi ${userFirstName}</p>
        <p>
          Thank you for signing up! Enter the code below in the application to finish verifying your email
          address.
        </p>
        <div class="code">${verificationCode}</div>
        <p>If you didn't create an account, please ignore this email.</p>
      </div>
      <div class="footer">
        <p>
          © ${new Date().getFullYear()} ${restaurantName}. All rights reserved.
        </p>
      </div>
    </div>
  </body>
</html>`;

const trackingOrderEmailTemplate = (restaurantName: string, customerName: string, trackingUrl: string) => `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Email Verification</title>
    <style>
      body {
        font-family: Arial, sans-serif;
        margin: 0;
        padding: 0;
      }
      .container {
        width: 100%;
        max-width: 30rem;
        margin: 0 auto;
        overflow: hidden;
      }
      .header {
        padding: 0.5rem;
        text-align: center;
      }
      .header h2 {
        color: #737373;
      }
      .content {
        padding: 0.5rem;
        line-height: 1.6;
        text-align: center;
        text-wrap: balance;
        color: #737373;
        font-size: 1rem;
      }
      .link {
        display: inline-block;
        padding: 10px 15px;
        background-color: black;
        border-radius: 5px;
        font-size: 16px;
        margin: 1rem 0;
        font-weight: 600;
        letter-spacing: 0.05rem;
        text-decoration: inherit;
        color: white;
        cursor: auto;
      }
      .footer {
        text-align: center;
        padding: 10px;
        font-size: 12px;
        color: #777;
        border-top: 1px solid #e0e0e0;
      }
    </style>
  </head>
  <body>
    <div class="container">
      <div class="header">
        <h2>${restaurantName}</h2>
        <h1>Verify your email</h1>
      </div>
      <div class="content">
        <p>Hi ${customerName}</p>
        <p>
          Thank you for placing an order with us! You can track your order by clicking the link below.
        </p>
        <a href=${trackingUrl} class="link">Track order</a>
      </div>
      <div class="footer">
        <p>
          © ${new Date().getFullYear()} ${restaurantName}. All rights reserved.
        </p>
      </div>
    </div>
  </body>
</html>`;

export { verificationEmailTemplate, trackingOrderEmailTemplate };
