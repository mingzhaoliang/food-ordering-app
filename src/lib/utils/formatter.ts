const stateConversion = (state: string) => {
  const states: { [key: string]: string } = {
    ACT: "Australian Capital Territory",
    NSW: "New South Wales",
    NT: "Northern Territory",
    QLD: "Queensland",
    SA: "South Australia",
    TAS: "Tasmania",
    VIC: "Victoria",
    WA: "Western Australia",
  };

  return states[state];
};

const reversedStateConversion = (state: string) => {
  const states: { [key: string]: string } = {
    "Australian Capital Territory": "ACT",
    "New South Wales": "NSW",
    "Northern Territory": "NT",
    Queensland: "QLD",
    "South Australia": "SA",
    Tasmania: "TAS",
    Victoria: "VIC",
    "Western Australia": "WA",
  };

  return states[state];
};

const shortPriceFormatter = (number: number) =>
  new Intl.NumberFormat("en-AU", {
    style: "currency",
    currency: "AUD",
    minimumFractionDigits: 0,
  }).format(number);

const priceFormatter = (number: number) =>
  new Intl.NumberFormat("en-AU", { style: "currency", currency: "AUD" }).format(number);

const timerFormatter = (duration: number) => {
  const seconds = Math.floor((duration / 1000) % 60);
  const minutes = Math.floor((duration / (1000 * 60)) % 60);
  const hours = Math.floor(duration / (1000 * 60 * 60));

  const formattedSeconds = seconds < 10 ? `0${seconds}` : `${seconds}`;
  const formattedMinutes = minutes < 10 ? `0${minutes}` : `${minutes}`;
  const formattedHours = hours === 0 ? "" : hours < 10 ? `0${hours}:` : `${hours}:`;
  const formattedDuration = `${formattedHours}${formattedMinutes}:${formattedSeconds}`;

  return formattedDuration;
};

export const datetimeFormatter = (date: Date) =>
  new Intl.DateTimeFormat("en-AU", { dateStyle: "medium", timeStyle: "short" }).format(date);

export { priceFormatter, reversedStateConversion, shortPriceFormatter, stateConversion, timerFormatter };
