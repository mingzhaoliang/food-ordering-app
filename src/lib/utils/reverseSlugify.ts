interface ReverseSlugifyOptions {
  delimiter?: string | RegExp; // Delimiter between words (e.g., '-', '_', or custom regex)
  caseStyle?: "none" | "capitalise" | "uppercase" | "lowercase" | "sentence"; // Handle case transformations
  trim?: boolean; // Whether to trim extra spaces
}

export const reverseSlugify = (slug: string, options?: ReverseSlugifyOptions): string => {
  const {
    delimiter = /[-_]/g, // Default to handling both hyphens and underscores
    caseStyle = "capitalise", // Default to capitalizing each word
    trim = true, // Trim extra spaces by default
  } = options || {};

  // Replace the delimiter with spaces
  let result = slug.replace(delimiter, " ");

  // Apply case transformations based on the caseStyle option
  switch (caseStyle) {
    case "capitalise":
      result = result.replace(/\b\w/g, (char) => char.toUpperCase()); // Capitalise first letter of each word
      break;
    case "uppercase":
      result = result.toUpperCase();
      break;
    case "lowercase":
      result = result.toLowerCase();
      break;
    case "sentence":
      // Capitalise only the first letter of the first word
      result = result.charAt(0).toUpperCase() + result.slice(1).toLowerCase();
      break;
    // 'none' case does nothing and retains the original case
  }

  // Optionally trim extra spaces
  if (trim) {
    result = result.trim();
  }

  return result;
};
