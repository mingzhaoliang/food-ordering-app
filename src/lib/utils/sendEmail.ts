import nodemailer from "nodemailer";

export const sendEmail = async (email: string, subject: string, content: string) => {
  const transporter = nodemailer.createTransport({
    service: "gmail",
    host: "smtp.gmail.com",
    secure: false,
    auth: {
      user: process.env.NM_GMAIL_FROM,
      pass: process.env.NM_GMAIL_APP_PASSWORD,
    },
  });

  try {
    await transporter.sendMail({
      from: process.env.NM_GMAIL_FROM,
      to: email,
      subject,
      html: content,
    });
  } catch (error) {
    console.error(error);

    throw new Error("Failed to send email");
  }
};
