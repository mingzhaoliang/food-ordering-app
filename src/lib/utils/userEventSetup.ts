import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

export const userEventSetup = (jsx: JSX.Element) => {
  return {
    user: userEvent.setup(),
    ...render(jsx),
  };
};
