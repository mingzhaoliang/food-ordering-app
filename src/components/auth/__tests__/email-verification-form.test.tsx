import { userEventSetup } from "@/lib/utils/userEventSetup";
import { fireEvent, render, screen } from "@testing-library/react";
import { redirect } from "next/navigation";
import EmailVerificationForm from "../email-verification-form";

vi.mock("@/actions/auth/verifyEvc");
vi.mock("@/actions/auth/verifyEmail");
vi.mock("next/navigation", () => ({
  redirect: vi.fn((url: string) => {
    console.log("Redirecting to: ", url);
  }),
}));

// Mock the ResizeObserver
const ResizeObserverMock = vi.fn(() => ({
  observe: vi.fn(),
  unobserve: vi.fn(),
  disconnect: vi.fn(),
}));

// Stub the global ResizeObserver
vi.stubGlobal("ResizeObserver", ResizeObserverMock);

describe("EmailVerificationForm", () => {
  it("should match the snapshot", () => {
    const { asFragment } = render(<EmailVerificationForm email="" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the form", () => {
    render(<EmailVerificationForm email="" />);
    expect(screen.getByRole("form")).toBeInTheDocument();
  });

  it("should render the buttons", () => {
    render(<EmailVerificationForm email="" />);
    expect(screen.getByRole("button", { name: "Verify" })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "Resend (60s)" })).toBeInTheDocument();
  });

  it("should render the error message for invalid input", async () => {
    const { user } = userEventSetup(<EmailVerificationForm email="" />);
    await user.click(screen.getByRole("button", { name: "Verify" }));
    expect(
      await screen.findByText("Your email verification code must be 6 alphanumeric characters.")
    ).toBeInTheDocument();
  });

  it("should render the error message for an incorrect email verification code", async () => {
    const { user } = userEventSetup(<EmailVerificationForm email="" />);
    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "123455" } });
    await user.click(screen.getByRole("button", { name: "Verify" }));
    expect(await screen.findByText("Invalid email verification code.")).toBeInTheDocument();
  });

  it("should redirect if verification succeeds", async () => {
    const { user } = userEventSetup(<EmailVerificationForm email="" />);
    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "123456" } });
    await user.click(screen.getByRole("button", { name: "Verify" }));
    expect(redirect).toHaveBeenCalledWith("/");
  });
});
