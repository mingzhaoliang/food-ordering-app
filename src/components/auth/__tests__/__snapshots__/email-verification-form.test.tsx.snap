// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmailVerificationForm > should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    class="space-y-8"
    role="form"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        Email Verification Code
      </label>
      <noscript />
      <div
        class="flex items-center gap-2 has-[:disabled]:opacity-50"
        data-input-otp-container="true"
        style="position: relative; cursor: text; user-select: none; pointer-events: none; --root-height: 0px;"
      >
        <div
          class="flex items-center"
        >
          <div
            class="relative flex h-10 w-10 items-center justify-center border-y border-r border-input text-sm transition-all first:rounded-l-md first:border-l last:rounded-r-md"
          />
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="relative flex h-10 w-10 items-center justify-center border-y border-r border-input text-sm transition-all first:rounded-l-md first:border-l last:rounded-r-md"
          />
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="relative flex h-10 w-10 items-center justify-center border-y border-r border-input text-sm transition-all first:rounded-l-md first:border-l last:rounded-r-md"
          />
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="relative flex h-10 w-10 items-center justify-center border-y border-r border-input text-sm transition-all first:rounded-l-md first:border-l last:rounded-r-md"
          />
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="relative flex h-10 w-10 items-center justify-center border-y border-r border-input text-sm transition-all first:rounded-l-md first:border-l last:rounded-r-md"
          />
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="relative flex h-10 w-10 items-center justify-center border-y border-r border-input text-sm transition-all first:rounded-l-md first:border-l last:rounded-r-md"
          />
        </div>
        <div
          style="position: absolute; inset: 0; pointer-events: none;"
        >
          <input
            aria-describedby=":r0:-form-item-description"
            aria-invalid="false"
            autocomplete="one-time-code"
            class="disabled:cursor-not-allowed"
            data-input-otp="true"
            id=":r0:-form-item"
            inputmode="numeric"
            maxlength="6"
            name="evc"
            pattern="^[a-zA-Z0-9]+$"
            style="position: absolute; inset: 0; width: 100%; height: 100%; display: flex; text-align: left; opacity: 1; color: transparent; pointer-events: all; background: transparent; caret-color: transparent; border: 0px solid transparent; outline: 0 solid transparent; box-shadow: none; line-height: 1; letter-spacing: -.5em; font-family: monospace; font-variant-numeric: tabular-nums;"
            value=""
          />
        </div>
      </div>
      <p
        class="text-sm text-muted-foreground"
        id=":r0:-form-item-description"
      >
        Please enter the code sent to your email:
        <br />
        <span
          class="underline"
        />
      </p>
    </div>
    <div
      class="space-x-3"
    >
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-full text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-active disabled:hover:bg-primary h-10 px-4 py-2 2xl:h-11"
        type="submit"
      >
        Verify
      </button>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-full text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 2xl:h-11"
        disabled=""
        type="button"
      >
        Resend (60s)
      </button>
    </div>
  </form>
</DocumentFragment>
`;
